### PR TITLE
Ensures that EVP encryption & decryption operations check the encrypt flag on the context.

### PR DESCRIPTION
**Pull request by alokmenghrajani**
I'm new to this codebase, so sorry if I made any style mistakes. Also, I have a couple questions:
1. do we need to set EVPerr before returning 0;
2. should we unittest this case?

I believe EVP should make it hard to shoot yourself in the foot, so this change ensures that a user cannot accidentally decrypt data with an encryption context or vice-versa. For example, without the check, if an encryption context is used to decrypt EVP_aes_256_gcm encrypted data, the code will fail to validate the TAG.

Example code availabe at: http://quaxio.com/wtf/openssl_wtf.html

### CODE REVIEW

**Comment by richsalz** [comment]
I like the concept, but I think returning 0 without setting an error code isn't a good idea.

**Comment by alokmenghrajani** [comment]
Feel free to modify the PR and merge it.

File: crypto/evp/evp_enc.c [hunk 1]
```diff
@@ -295,7 +295,7 @@ int EVP_DecryptInit_ex(EVP_CIPHER_CTX *ctx, const EVP_CIPHER *cipher, ENGINE *im
 	return EVP_CipherInit_ex(ctx, cipher, impl, key, iv, 0);
 	}
 
-int EVP_EncryptUpdate(EVP_CIPHER_CTX *ctx, unsigned char *out, int *outl,
+int EVP_EncryptOrDecryptUpdate(EVP_CIPHER_CTX *ctx, unsigned char *out, int *outl,
```
Comment (levitte) [inline crypto/evp/evp_enc.c]: This should be made static and with lowercase prefix (i.e. `evp_` rather than `EVP_`

**Comment by levitte** [comment]
It's unfortunate that this hasn't been looked at for so long.  The code looks correct (apart from the nit I picked), but it needs a rebase to a fresher master.

**Comment by mattcaswell** [comment]
This PR is from pre-reformat times. Either we should bring it up to date and fix the nit, or abandon it. @levitte?

**Comment by levitte** [comment]
Like I said, "it needs a rebase to a fresher master"

**Comment by mattcaswell** [comment]
Ok. So I suggest unless @alokmenghrajani feels like bringing this up to date, we should close this PR.

**Comment by levitte** [comment]
Agreed

**Comment by alokmenghrajani** [comment]
Not psyched to go over this ~2 years later.

**Comment by owlstead** [comment]
That must be the most fucked up reason to close an issue *ever*.

See for instance the question on SO in the next URL to see that this causes issues.

https://stackoverflow.com/questions/53690728/openssl-evp-decryption-fails-for-ecb-and-cbc-but-works-for-ofb/53691697#comment94239271_53691697

The routines are currently ignoring the fail fast and least surprise design principles. Please fix.

**Comment by kroeckx** [comment]
Feel free to create an other pull request to add those checks.

**Comment by alokmenghrajani** [comment]
@owlstead re "Please don't close issues if they still persist.", this wasn't an issue but a PR. By my books, it's fine to close PRs if no progress is being made.

@levitte Appreciate that you eventually got to fix this issue.

**Comment by owlstead** [comment]
@alokmenghrajani Thank you for initially proposing the code. But your PR consists of code to fix an issue, correct? It doesn't offer any other functionality but to correctly report an error if the API is misused.

**Comment by alokmenghrajani** [comment]
@owlstead thank you for eventually applying the PR and not keeping the author from the original commit. Sure. Happy?

**Comment by owlstead** [comment]
Hmm, not about the author part, but you're clearly the initiator and it is fixed so yeah - very happy :)

**Pull request closed without merging**